File: week 2/Problem Set 2/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);
int reading_level(int letter, int word, int sentence);

int main(void)
{
    string input = get_string("Text: ");
    int grade = reading_level(count_letters(input), count_words(input), count_sentences(input));
    if (grade >= 16)
    {
        printf("Grade 16+\n");
    }
    else if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade %i\n", grade);
    }
}

// Count the number of letters in the text
int count_letters(string text)
{
    int count = 0;
    for (int i = 0, l = strlen(text); i < l; i++)
    {
        char current = toupper(text[i]);
        if (current >= 65 && current <= 90)
        {
            count++;
        }
    }
    return count;
}

// Count the number of words in the text
int count_words(string text)
{
    int count = 1;
    for (int i = 0, l = strlen(text); i < l; i++)
    {
        char current = toupper(text[i]);
        if (current == 32)
        {
            count++;
        }
    }
    return count;
}

// Count the number of sentences in the text
int count_sentences(string text)
{
    int count = 0;
    for (int i = 0, l = strlen(text); i < l; i++)
    {
        char current = toupper(text[i]);
        if (current == 33 || current == 46 || current == 63)
        {
            count++;
        }
    }
    return count;
}

// Calculate the reading grade
int reading_level(int letter, int word, int sentence)
{
    float letters_per_100 = (float) letter / (float) word * 100;
    float sentences_per_100 = (float) sentence / (float) word * 100;
    float index = 0.0588 * letters_per_100 - 0.296 * sentences_per_100 - 15.8;
    return round(index);
}

File: week 6/Problem Set 6/readability.py
def main():
    text = input("Text: ")
    letters = count_letters(text)
    words = len(text.split(" "))
    sentences = count_sentences(text)
    average_letters = letters / words * 100
    average_sentence = sentences / words * 100
    index = 0.0588 * average_letters - 0.296 * average_sentence - 15.8

    if index >= 16:
        print("Grade 16+")
    elif index < 1:
        print("Before Grade 1")
    else:
        print(f"Grade {round(index)}")


def count_letters(text):
    count = 0
    for letter in text:
        if letter.isalpha():
            count += 1
    return count


def count_sentences(text):
    count = 0
    for letter in text:
        if letter == "!" or letter == "." or letter == "?":
            count += 1
    return count


if __name__ == "__main__":
    main()
